Support for Rails 7.1 (#46)
Shopify/ejson-rails: README.md

will be accessible via `Rails.application.secrets.some_secret` or `Rails.application.secrets[:some_secret]` on boot. JSON files are loaded once and contents are `deep_merge`'d into your app's existing rails secrets.

Secrets will also be accessible via `Rails.application.credentials`, e.g. `Rails.application.credentials.some_secret` or `Rails.application.credentials[:some_secret]`. To avoid subtle compatibility issues, if a credential already exists, an error will occur.

NOTE: This gem does not decrypt ejson for you. You will need to configure this as part of your deployment pipeline.

> ## Migrating to credentials

Reviewer: I suppose because of the human nature or ERB, and ambiguity like that, we're not going to be able to make a migration script then

Rails 7.1 has deprecated application secrets in favor of credentials. ejson-rails can merge the secrets to application credentials to allow for a migration.

Even before running Rails 7.1, you can migrate your secrets in a few steps.

First, move the development and test secrets to JSON secrets:

```sh-session
bin/rails runner 'Rails.root.join("config/secrets.#{Rails.env}.json").write(JSON.pretty_generate(Rails.application.secrets.to_h.without(:secret_key_base)))'
bin/rails runner -e test 'Rails.root.join("config/secrets.#{Rails.env}.json").write(JSON.pretty_generate(Rails.application.secrets.to_h.without(:secret_key_base)))'
```

> Secrets support ERB while EJSON secrets don't, so if you used that feature, you will need to move that logic in the environment configurations:

Reviewer 2: ```suggestion
Secrets support ERB while EJSON secrets don't, so if your secrets contain ERB, you will need to move that logic to the environment configurations:
```

**Before**:

`config/secrets.yml`
```yaml
development:
  some_external_service:
    api_token: <%= ENV.fetch(SOME_EXTERNAL_SERVICE_API_TOKEN, "12345") %>
```

**After**:

`config/secrets.development.json` as generated by the command above.
```json
{
  "some_external_service": {
    "api_token": "12345"
  },
  "something_else_entirely": "abc"
}
```

`config/environments/development.rb`
```ruby
Rails.application.configure do
  # elided

  # credential should be set using []=, not the dynamic accessors
  credentials[:some_external_service][:api_token] = ENV.fetch("SOME_EXTERNAL_SERVICE_API_TOKEN", "12345")

  # top-level values must be set through `credentials.config`
  credentials.config[:something_else_entirely] = ENV.fetch("SOMETHING_ELSE_ENTIRELY", "abc")
end
```

> You could remove the default value from the JSON or instead only set the value if the environment variable is available.

Reviewer 2: I don't understand the "You could remove the default value from the JSON" part 😅 Could we have a code example to go along with this?

Author: Right, by simply dumping secrets.yml into secrets.#{env}.json, we got the default values from the ENV, "12345". 
You get:
```json
{ "some": { "secret": "12345" } }
```
and 
```ruby
credentials[:some][:secret] = ENV.fetch("FOO", "12345")
```
so the dummy value is duplicated.

Instead either:
* you keep the dummy in the json, and you change the environment to be:
```ruby
if foo = ENV["FOO"]
  credentials[:some][:secret] = foo
end
```
* you keep the dummy in the environment, and you change the json to be
```json
{ "some": { "secret": null }
```

Maybe we can remove that sentence entirely and let people handle the duplication as they see fit.

```ruby
Rails.application.credentials.some_external_service.api_token = "foo"
Rails.application.credentials[:some_external_service][:api_token] # => "12345"
```

Also note the code sets top-level values through `credentials.config`, because `credentials#[]=(key, value)` sets values in a different object.

```ruby
Rails.application.credentials[:something_else_entirely] = "foo"
Rails.application.credentials[:something_else_entirely] # => "abc"
```

Make sure there's no code using the dynamic accessors before setting the configuration in the Hash, or the values won't be accessible from the dynamic accessor:

```ruby
Rails.application.credentials.something_else_entirely # just accessing is enough to cause the issue
Rails.application.credentials[:some_external_service][:api_token] = "foo"
Rails.application.credentials.some_external_service.api_token # => "12345"
```

You're now ready to use credentials instead of secrets:

```sh-session
git ls-files | xargs ruby -pi -e 'gsub("Rails.application.secrets", "Rails.application.credentials")' --
```

> Once you're running Rails 7.1, to avoid the deprecation warning from the use of secrets in `ejson-rails`, you can require another file from your Gemfile:

Reviewer 2: ```suggestion
To avoid the deprecation warning from the use of secrets in `ejson-rails` once you're running Rails 7.1, require another file from your Gemfile:
```
